File: DFRobot_VEML6075.cpp
#include <DFRobot_VEML6075.h>

#define UVA_A_COEF    2.22f
#define UVA_B_COEF    1.33f
#define UVA_C_COEF    2.95f
#define UVA_D_COEF    1.74f

#define UV_ALPHA      1.0
#define UV_BETA       1.0
#define UV_GAMMA      1.0
#define UV_DELTA      1.0

#define UVA_RESPONSIVITY_100MS    0.001111
#define UVB_RESPONSIVITY_100MS    0.00125

const float UvaResponsivityList[5] = {
  UVA_RESPONSIVITY_100MS / 0.5016286645, // 50ms
  UVA_RESPONSIVITY_100MS,                // 100ms
  UVA_RESPONSIVITY_100MS / 2.039087948,  // 200ms
  UVA_RESPONSIVITY_100MS / 3.781758958,  // 400ms
  UVA_RESPONSIVITY_100MS / 7.371335505   // 800ms
};

const float UvbResponsivityList[5] = {
  UVB_RESPONSIVITY_100MS / 0.5016286645, // 50ms
  UVB_RESPONSIVITY_100MS,                // 100ms
  UVB_RESPONSIVITY_100MS / 2.039087948,  // 200ms
  UVB_RESPONSIVITY_100MS / 3.781758958,  // 400ms
  UVB_RESPONSIVITY_100MS / 7.371335505   // 800ms
};

boolean DFRobot_VEML6075::begin()
{
  uint16_t   tmp = 0;
  readReg(VEML6075_ID, (uint8_t*) &tmp, sizeof(tmp));
  if(tmp == VEML6075_ID_DEFAULT) {
    tmp = VEML6075_CONF_DEFAULT;
    writeReg(VEML6075_CONF, (uint8_t*) &tmp, sizeof(tmp));
    setPower(eVEML6075PowerOn);
    setActiveForceMode(eVEML6075ActiveForceModeDisable);
    setIntegrationTime(eVEML6075UV_IT_100);
    return true;
  }
  return false;
}

#define writeRegBits(reg, buf, bits, value) \
  readReg((reg), (uint8_t*) &buf, sizeof(buf)); \
  (bits) = (value); \
  writeReg((reg), (uint8_t*) &buf, sizeof(buf))

// main class config functions --------------------------------------------------------------------------

void DFRobot_VEML6075::setPower(eVEML6075Power_t ePower)
{
  sVEML6075Conf_t   sConf;
  writeRegBits(VEML6075_CONF, sConf, sConf.SD, ePower);
}

void DFRobot_VEML6075::setActiveForceMode(eVEML6075ActiveForceMode_t eMode)
{
  sVEML6075Conf_t   sConf;
  writeRegBits(VEML6075_CONF, sConf, sConf.UV_AF, eMode);
  _eActiveForceMode = eMode;
}

void DFRobot_VEML6075::setDynamic(eVEML6075Dynamic_t eDynamic)
{
  sVEML6075Conf_t   sConf;
  writeRegBits(VEML6075_CONF, sConf, sConf.HD, eDynamic);
}

void DFRobot_VEML6075::setIntegrationTime(eVEML6075UV_IT_t eUV_IT)
{
  sVEML6075Conf_t   sConf;
  writeRegBits(VEML6075_CONF, sConf, sConf.UV_IT, eUV_IT);
  _eUV_IT = eUV_IT;
}

void DFRobot_VEML6075::trigOneMeasurement()
{
  if(_eActiveForceMode == eVEML6075ActiveForceModeEnable) {
    sVEML6075Conf_t   sConf;
    writeRegBits(VEML6075_CONF, sConf, sConf.UV_TRIG, VEML6075_TRIG_ONE_MEASUREMENT);
  }
}

eVEML6075ActiveForceMode_t DFRobot_VEML6075::getActiveForceMode()
{
  sVEML6075Conf_t   sConf;
  readReg(VEML6075_CONF, (uint8_t*) &sConf, sizeof(sConf));
  return (eVEML6075ActiveForceMode_t) sConf.UV_AF;
}

eVEML6075Dynamic_t DFRobot_VEML6075::getDynamic()
{
  sVEML6075Conf_t   sConf;
  readReg(VEML6075_CONF, (uint8_t*) &sConf, sizeof(sConf));
  return (eVEML6075Dynamic_t) sConf.HD;
}

eVEML6075Power_t DFRobot_VEML6075::getPower()
{
  sVEML6075Conf_t   sConf;
  readReg(VEML6075_CONF, (uint8_t*) &sConf, sizeof(sConf));
  return (eVEML6075Power_t) sConf.SD;
}

eVEML6075UV_IT_t DFRobot_VEML6075::getIntegrationTime()
{
  sVEML6075Conf_t   sConf;
  readReg(VEML6075_CONF, (uint8_t*) &sConf, sizeof(sConf));
  return (eVEML6075UV_IT_t) sConf.UV_IT;
}

// main class process functions ------------------------------------------------------------------------------------

uint16_t DFRobot_VEML6075::readUvComp1Raw()
{
  uint16_t    tmp;
  readReg(VEML6075_UV_COMP1, (uint8_t*) &tmp, sizeof(tmp));
  return tmp;
}

uint16_t DFRobot_VEML6075::readUvComp2Raw()
{
  uint16_t    tmp;
  readReg(VEML6075_UV_COMP2, (uint8_t*) &tmp, sizeof(tmp));
  return tmp;
}

uint16_t DFRobot_VEML6075::readUvaRaw()
{
  uint16_t    tmp;
  readReg(VEML6075_UVA, (uint8_t*) &tmp, sizeof(tmp));
  return tmp;
}

uint16_t DFRobot_VEML6075::readUvbRaw()
{
  uint16_t    tmp;
  readReg(VEML6075_UVB, (uint8_t*) &tmp, sizeof(tmp));
  return tmp;
}

float DFRobot_VEML6075::getUva()
{
  return ((float) readUvaRaw() - ((UVA_A_COEF * UV_ALPHA * (float) readUvComp1Raw()) / UV_GAMMA)
                               - ((UVA_B_COEF * UV_ALPHA * (float) readUvComp2Raw()) / UV_DELTA));
}

float DFRobot_VEML6075::getUvb()
{
  return ((float) readUvbRaw() - ((UVA_C_COEF * UV_BETA * (float) readUvComp1Raw()) / UV_GAMMA)
                               - ((UVA_D_COEF * UV_BETA * (float) readUvComp2Raw()) / UV_DELTA));
}

float DFRobot_VEML6075::getUvi(float Uva, float Uvb)
{
  Uva = (Uva * (1.0f / UV_ALPHA) * UvaResponsivityList[(uint8_t) _eUV_IT]);
  Uvb = (Uvb * (1.0f / UV_BETA) * UvbResponsivityList[(uint8_t) _eUV_IT]);
  return (Uva + Uvb) / 2.0f;
}

// for DFRobot_VEML6075_IIC ---------------------------------------------------------------------------------------------

void DFRobot_VEML6075_IIC::writeReg(uint8_t reg, uint8_t *pBuf, uint16_t len)
{
  _pWire->begin();
  _pWire->beginTransmission(_addr);
  _pWire->write(reg);
  for(uint16_t i = 0; i < len; i ++)
    _pWire->write(pBuf[i]);
  _pWire->endTransmission();
}

void DFRobot_VEML6075_IIC::readReg(uint8_t reg, uint8_t *pBuf, uint16_t len)
{
  _pWire->begin();
  _pWire->beginTransmission(_addr);
  _pWire->write(reg);
  if(_pWire->endTransmission(false) != 0)
    return;
  _pWire->requestFrom(_addr, (uint8_t) len);
  for(uint16_t i = 0; i < len; i ++)
    pBuf[i] = _pWire->read();
  _pWire->endTransmission();
}

File: python/raspberrypi/DFRobot_VEML6075.py
# -*- coding: utf-8 -*-
'''!
  @file DFRobot_VEML6075.py
  @brief define DFRobot_VEML6075 class infrastructure, the implementation of basic methods
  @copyright	Copyright (c) 2010 DFRobot Co.Ltd (http://www.dfrobot.com)
  @license     The MIT License (MIT)
  @maintainer [Fary](feng.yang@dfrobot.com)
  @version  V1.0
  @date  2021-10-18
  @url https://github.com/DFRobot/DFRobot_VEML6075
'''
from raspberry_iic import RaspberryIIC

class DFRobot_VEML6075:
  '''!
    @brief DFRobot_VEML6075
  '''
  ID_DEFAULT = 0x26

  CONF = 0x00
  CONF_DEFAULT = 0x00
  UVA = 0x07
  UVB = 0x09
  UV_COMP1 = 0x0a
  UV_COMP2 = 0x0b
  ID = 0x0c

  UVA_RESPONSIVITY_100MS = 0.001111
  UVB_RESPONSIVITY_100MS = 0.00125

  uva_responsivity_list = [
    UVA_RESPONSIVITY_100MS / 0.5016286645,
    UVA_RESPONSIVITY_100MS,               
    UVA_RESPONSIVITY_100MS / 2.039087948, 
    UVA_RESPONSIVITY_100MS / 3.781758958, 
    UVA_RESPONSIVITY_100MS / 7.371335505  
  ]

  Uvb_responsivity_list = [
    UVB_RESPONSIVITY_100MS / 0.5016286645,
    UVB_RESPONSIVITY_100MS,               
    UVB_RESPONSIVITY_100MS / 2.039087948, 
    UVB_RESPONSIVITY_100MS / 3.781758958, 
    UVB_RESPONSIVITY_100MS / 7.371335505  
  ]

  UV_IT_50 = 0x00
  UV_IT_100 = 0x10
  UV_IT_200 = 0x20
  UV_IT_400 = 0x30
  UV_IT_800 = 0x40

  POWER_ON = 0x00
  POWER_OFF = 0x01

  ACTIVE_FORCE_ENABLE = 0x02
  ACTIVE_FORCE_DISABLE = 0x00

  DYNAMIC_NORMAL = 0x00
  DYNAMIC_HIGH = 0x08

  def __init__(self, i2cBus, addr):
    '''!
      @brief   _init_
      @param i2cBus Select _bus
      @param addr Module address
    '''
    self._bus = RaspberryIIC(i2cBus)
    self._addr = addr
    self._is_active_force_mode = False
    self._UV_IT = self.UV_IT_100 >> 4

  def begin(self):
    '''!
      @brief initialization function
      @return bool,returns the initialization status
      @retval True Initialization succeeded
      @retval False Initialization  failed
    '''
    _id = self._read_reg(self.ID, 2)
    if _id[0] == self.ID_DEFAULT:
      self._write_reg(self.CONF, [self.CONF_DEFAULT & 0xff, self.CONF_DEFAULT >> 8])
      self.set_active_force_mode(self.ACTIVE_FORCE_DISABLE)
      self.set_power(self.POWER_ON)
      self.set_integration_time(self.UV_IT_100)
      return True
    return False

  def _write_reg(self, reg, val):
    '''!
      @brief write the sensor IIC data
      @param reg register
      @param val write the store and buffer of the data
      @return return the actually written length
    '''
    return self._bus.write_reg(self._addr, reg, val)

  def _read_reg(self, reg, len):
    '''!
      @brief get the sensor IIC data
      @param reg register
      @param len data length to be readed 
      @return return the actually readed length
    '''
    return self._bus.read_reg(self._addr, reg, len)

  def _write_reg_bits(self, reg, field, val):
    '''!
      @brief Sets bits of a register
      @param reg register
      @param field field
      @param val Value to be set
    '''
    buf = self._read_reg(reg, 2)
    buf[0] &= field
    buf[0] |= val
    self._write_reg(reg, buf)

  def set_power(self, power):
    '''!
      @brief Set power
      @param power Power status to set
    '''
    self._write_reg_bits(self.CONF, 0xfe, power)

  def uvi2mwpcm2(self, Uvi):
    '''!
      @brief Data conversion
      @return mW/cm2
    '''
    return (Uvi * 0.0025)

  def get_power(self):
    '''!
      @brief get power status
      @return  Power status
      @retval    1    power on
      @retval    0   power off
    '''
    buf = self._read_reg(self.CONF, 2)
    return buf[0] & 0x01

  def set_active_force_mode(self, isEnable):
    '''!
      @brief set active force mode
      @param isEnable 
      @n       ACTIVE_FORCE_ENABLE    VEML6075 active force mode enable
      @n       ACTIVE_FORCE_DISABLE   VEML6075 active force mode disable
    '''
    self._write_reg_bits(self.CONF, 0xfd, isEnable)
    if isEnable == self.ACTIVE_FORCE_ENABLE:
      self._is_active_force_mode = True

  def get_active_force_mode(self):
    '''!
      @brief get active force mode
      @return  mode 
      @retval    ACTIVE_FORCE_ENABLE    VEML6075 active force mode enable
      @retval    ACTIVE_FORCE_DISABLE   VEML6075 active force mode disable
    '''
    buf = self._read_reg(self.CONF, 2)
    return buf[0] & 0x02

  def trig_one_measurement(self):
    '''!
      @brief  Trigger a measurement
    '''
    if self._is_active_force_mode:
      self._write_reg_bits(self.CONF, 0xfb, 0x04)

  def set_dynamic(self, dynamic):
    '''!
      @brief set Dynamic
      @param dynamic
      @n       DYNAMIC_NORMAL    normal dynamic setting
      @n       DYNAMIC_HIGH      high dynamic setting
    '''
    self._write_reg_bits(self.CONF, 0xf7, dynamic)

  def get_dynamic(self):
    '''!
      @brief get Dynamic
      @return  dynamic
      @n       DYNAMIC_NORMAL    normal dynamic setting
      @n       DYNAMIC_HIGH      high dynamic setting
    '''
    buf = self._read_reg(self.CONF, 2)
    return buf[0] & 0x08

  def set_integration_time(self, t):
    '''!
      @brief set integration time
      @param   t
      @n       UV_IT_50     50ms
      @n       UV_IT_100    100ms
      @n       UV_IT_200    200ms
      @n       UV_IT_400    400ms
      @n       UV_IT_800    800ms
    '''
    self._write_reg_bits(self.CONF, 0x8f, t)
    self._UV_IT = t >> 4

  def get_integration_time(self):
    '''!
      @brief get integration time
      @return  integration time
      @retval    UV_IT_50     50ms
      @retval    UV_IT_100    100ms
      @retval    UV_IT_200    200ms
      @retval    UV_IT_400    400ms
      @retval    UV_IT_800    800ms
    '''
    buf = self._read_reg(self.CONF, 2)
    return buf[0] & 0x8f

  def read_uva_raw(self):
    '''!
      @brief The reading sensor collects raw data in the UVA band
      @return  UVA Raw
    '''
    buf = self._read_reg(self.UVA, 2)
    return buf[0] | (buf[1] << 8)

  def read_uvb_raw(self):
    '''!
      @brief The reading sensor collects raw data in the UVB band
      @return  UVB Raw
    '''
    buf = self._read_reg(self.UVB, 2)
    return buf[0] | (buf[1] << 8)

  def read_UVcomp1_raw(self):
    '''!
      @brief The reading sensor collects raw data in the UvComp1
      @return  UvComp1 Raw
    '''
    buf = self._read_reg(self.UV_COMP1, 2)
    return buf[0] | (buf[1] << 8)
  
  def read_UVcomp2_raw(self):
    '''!
      @brief The reading sensor collects raw data in the UvComp2
      @return  UvComp2 Raw
    '''
    buf = self._read_reg(self.UV_COMP2, 2)
    return buf[0] | (buf[1] << 8)

  def get_uva(self):
    '''!
      @brief The UVA band data of ultraviolet ray are calculated according to the data collected
      @return  UVA data
    '''
    return (self.read_uva_raw() - ((2.22 * 1.0 * self.read_UVcomp1_raw()) / 1.0) - ((1.33 * 1.0 * self.read_UVcomp2_raw()) / 1.0))

  def get_uvb(self):
    '''!
      @brief The UVB band data of ultraviolet ray are calculated according to the data collected
      @return  UVB data
    '''
    return (self.read_uvb_raw() - ((2.95 * 1.0 * self.read_UVcomp1_raw()) / 1.0) - ((1.74 * 1.0 * self.read_UVcomp2_raw()) / 1.0))

  def get_uvi(self, Uva, Uvb):
    '''!
      @brief The UVI band data of ultraviolet ray are calculated according to the data collected
      @return  UVI data
    '''
    Uva = ((Uva * (1.0 / 1.0)) * self.uva_responsivity_list[self._UV_IT])
    Uvb = ((Uvb * (1.0 / 1.0)) * self.uva_responsivity_list[self._UV_IT])
    return (Uva + Uvb) / 2
